README
= BigSitemap

== DESCRIPTION:

BigSitemap is a Sitemap generator specifically designed for large sites (although it works equally well with small sites).  It splits large Sitemaps into multiple files, gzips the files to minimize bandwidth usage, batches database queries so it doesn't take your site down, can be set up with just a few lines of code and is compatible with just about any framework.


== INSTALL:

* Via git: git clone git://github.com/alexrabarts/big_sitemap.git
* Via gem: gem install alexrabarts-big_sitemap -s http://gems.github.com

== SYNOPSIS:

The minimum required to generated a sitemap is:

  sitemap = BigSitemap.new(:base_url => 'http://example.com')
  sitemap.add(:model => MyModel, :path => 'my_controller')
  sitemap.generate

You can put this in a rake/thor task and create a cron job to run it periodically.  It should be enough for most Rails/Merb applications.

Your models must provide either a <code>find_for_sitemap</code> or <code>all</code> class method that returns the instances that are to be included in the sitemap.  Additionally, you models must provide a <code>count_for_sitemap</code> or <code>count</code> class method that returns a count of the instances to be included.  If you're using ActiveRecord (Rails) or DataMapper then <code>all</code> and <code>count</code> are already provided and you don't need to do anything unless you want to include a subset of records.  If you provide your own <code>find_for_sitemap</code> or <code>all</code> method then it should be able to handle the <code>:offset</code> and <code>:limit</code> options, in the same way that ActiveRecord and DataMapper handle them.  This is especially important if you have more than 50,000 URLs.

To generate the URLs, BigSitemap will combine the constructor arguments with the <code>to_param</code> method of each instance returned (provided by ActiveRecord but not DataMapper).  If this method is not present, <code>id</code> will be used.  The URL is constructed as:

  ":base_url/:path/:to_param" # (if to_param exists)
  ":base_url/:path/:id"       # (if to_param does not exist)

BigSitemap knows about the document root of Rails and Merb.  If you are using another framework then you can specify the document root with the <code>:document_root</code> option.  e.g.:

  BigSitemap.new(:base_url => 'http://example.com', :document_root => "#{FOO_ROOT}/httpdocs")

By default, the sitemap files are created under <code>/sitemaps</code>.  You can modify this with the <code>:path</code> option:

  BigSitemap.new(:base_url => 'http://example.com', :path => 'google-sitemaps') # places Sitemaps under /google-sitemaps

Sitemaps will be split across several files if more than 50,000 records are returned.  You can customize this limit with the <code>:max_per_sitemap</code> option:

  BigSitemap.new(:base_url => 'http://example.com', :max_per_sitemap => 1000) # Max of 1000 URLs per Sitemap

The database is queries in batches to prevent large SQL select statements from locking the database for too long.  By default, the batch size is 1001 (not 1000 due to an obscure bug in DataMapper that appears when an offset of 37000 is used).  You can customize the batch size with the <code>:batch_size</code> option:

  BigSitemap.new(:base_url => 'http://example.com, :batch_size => 5000) # Database is queried in batches of 5,000

Google, Yahoo!, MSN and Ask are pinged once the Sitemap files are generated.  You can turn one or more of these off:

  BigSitemap.new(
    :base_url => 'http://example.com',
    :ping_google => false,
    :ping_yahoo  => false,
    :ping_msn    => false,
    :ping_ask    => false
  )

You must provide an App ID in order to ping Yahoo! (more info at http://developer.yahoo.com/search/siteexplorer/V1/updateNotification.html):

  BigSitemap.new(:base_url => 'http://example.com', :yahoo_app_id => 'myYahooAppId') # Yahoo! will now be pinged

== LIMITATIONS:

If your database is likely to shrink during the time it takes to create the sitemap then you might run into problems (the final, batched SQL select will overrun by setting a limit that is too large since it is calculated from the count, which is queried at the very beginning).  Patches welcome!

== TODO

* Support for priority and changefreq (currently hard-coded to 'weekly')

== CREDITS

Thanks to Alastair Brunton and Harry Love, who's work provided a starting point for this library.
http://scoop.cheerfactory.co.uk/2008/02/26/google-sitemap-generator/

== COPYRIGHT

Copyright (c) 2009 Stateless Systems (http://statelesssystems.com). See LICENSE for details.

History.txt
=== 1.1.0 / 2013-02-06

* Yandex is now also pinged when the sitemap is updated
* Deprecated methods from v1.0.0 now removed

=== 1.0.2 / 2013-02-06

* Fixes an issue with duplicate URLs

=== 1.0.1 / 2012-03-10

* Fixes an issue generating correct URL paths

=== 1.0.0 / 2011-10-24

* API Change: Sitemaps are now generated using a block syntax.  Find methods are no longer the responsibility of BigSitemap.  Instead, sitemaps are generated using a block, in which you call your own find methods, passing the results to BigSitemap with the 'add' method.  See the README for details.
* BigSitemapRails and BigSitemapMerb are now BigSitemap::Rails and BigSitemap::Merb, respectively.
* Sitemap files are now placed in the document root by default
* Sitemaps are now automatically cleaned before generating the new set
* Search engines are now pinged automatically when the sitemap is generated
* Lock files are now generated automatically
* Sitemap files are no longer split amongst your models

=== 0.8.5 / 2011-10-20

* Gzipped files now include indents and newlines

=== 0.8.4 / 2011-10-20

* Fixes an issue where joins where causing ambiguous "id" column
  (https://github.com/alexrabarts/big_sitemap/pull/17)
* Fixes an issue with empty <loc> nodes
  (https://github.com/alexrabarts/big_sitemap/pull/20)

=== 0.8.3 / 2011-03-08

* Separate URL and file paths are now supported via the :document_path
  and :url_path options
* Fixes an issue when initializing in Rails 3

=== 0.8.2 / 2011-01-25

* Fixes an issue where sitemap files were not being generated if the same model
  was added more than once (fixes issue #5: https://github.com/alexrabarts/big_sitemap/issues/#issue/5)

=== 0.8.1 / 2011-01-25

* API change: Rails/Merb are no longer automatically detected - use BigSitemapRails and BigSitemapMerb instead
* API change: Rails' polymorphic_url helper is no longer used to generate URLs (use a lambda with the new :location option instead)
* Static resources can now be added using the add_static method
* Incremental updates are now available via the :partial_update option
* "loc" URL values can now be generated with lambdas
* Sitemap files can now be locked while being generated using the with_lock method
* Several bug fixes

=== 0.5.1 / 2009-09-07

* Fixes an issue with the :last_modified key being passed into the find method options

=== 0.5.0 / 2009-09-07

* Add support for lambdas when specifying lastmod

=== 0.4.0 / 2009-08-09

* Use Bing instead of Live/MSN.  Note, this breaks backwards compatibility as
  the old :ping_msn option is now :ping_bing.

=== 0.3.5 / 2009-08-05

* Fixed bugs in root_url generation and url_for_sitemap generation

=== 0.3.4 / 2009-07-02

* BigSitemap-specific options are no longer passed through to the ORM's find method

=== 0.3.2 / 2009-06-09

* Better handling of URLs when Rails' polymorphic_url isn't available in the model

=== 0.3.2 / 2009-06-09

* Fixes "uninitialized constant ActionController" error
* Fixes "Unknown key(s): path" error

=== 0.3.1 / 2009-04-18

* Fixes broken gemspec

=== 0.3.0 / 2009-04-06

* API change: Pass model through as first argument to add method, e.g.sitemap.add(Posts, {:path => 'articles'})
* API change: Use Rails' polymorphic_url helper to generate URLs if Rails is being used
* API change: Only ping search engines when ping_search_engines is explicitly called
* Add support for passing options through to the model's find method, e.g. :conditions
* Allow base URL to be specified as a hash as well as a string
* Add support for changefreq and priority
* Pluralize sitemap model filenames
* GZipping may optionally be turned off

=== 0.2.1 / 2009-03-12

* Normalize path arguments so it no longer matters whether a leading slash is used or not

=== 0.2.0 / 2009-03-11

* Methods are now chainable

=== 0.1.4 / 2009-03-11

* Add clean method to clear out Sitemaps directory
* Make methods chainable

=== 0.1.3 / 2009-03-10

* Initial release



test/test_helper.rb
require 'rubygems'
require 'bundler/setup'

require 'test/unit'
require 'shoulda'
require 'mocha'
require 'test/fixtures/test_model'

require 'big_sitemap'

class Test::Unit::TestCase

  #TestHelper
  private
  def generate_sitemap(options={}, &block)
    BigSitemap.generate(options.merge(:base_url => 'http://example.com', :document_root => tmp_dir), &block)
  end

  def delete_tmp_files
    Dir["#{sitemaps_dir}/sitemap*"].each do |f|
      FileUtils.rm_rf f
    end
  end

  def sitemaps_index_file
    "#{unzipped_sitemaps_index_file}.gz"
  end

  def unzipped_sitemaps_index_file
    "#{sitemaps_dir}/sitemap_index.xml"
  end

  def unzipped_first_sitemap_file
    "#{sitemaps_dir}/sitemap.xml"
  end

  def first_sitemap_file
    "#{sitemaps_dir}/sitemap.xml.gz"
  end

  def second_sitemap_file
    "#{sitemaps_dir}/sitemap_1.xml.gz"
  end

  def third_sitemap_file
    "#{sitemaps_dir}/sitemap_2.xml.gz"
  end

  def sitemaps_dir
    tmp_dir
  end

  def tmp_dir
    '/tmp'
  end

  def ns
    {'s' => 'http://www.sitemaps.org/schemas/sitemap/0.9', 'v' => "http://www.google.com/schemas/sitemap-video/1.1"}
  end

  def elements(filename, el)
    file_class = filename.include?('.gz') ? Zlib::GzipReader : File
    data = Nokogiri::XML.parse(file_class.open(filename).read)
    data.search("//s:#{el}", "//v:#{el}", ns)
  end

  def num_elements(filename, el)
    elements(filename, el).size
  end
end


test/big_sitemap_test.rb
require File.dirname(__FILE__) + '/test_helper'
require 'nokogiri'

class BigSitemapTest < Test::Unit::TestCase
  def setup
    delete_tmp_files
  end

  def teardown
    delete_tmp_files
  end

  should 'raise an error if the :base_url option is not specified' do
    assert_nothing_raised { BigSitemap.new(:base_url => 'http://example.com', :document_root => tmp_dir) }
    assert_raise(ArgumentError) { BigSitemap.new(:document_root => tmp_dir) }
  end

  should 'generate the same base URL with :base_url option' do
    options = {:document_root => tmp_dir}
    url = 'http://example.com'
    sitemap = BigSitemap.new(options.merge(:base_url => url))

    assert_equal url, sitemap.instance_variable_get(:@options)[:base_url]
  end

  should 'generate the same base URL with :url_options option' do
    options = {:document_root => tmp_dir}
    url = 'http://example.com'
    sitemap = BigSitemap.new(options.merge(:url_options => {:host => 'example.com'}))

    assert_equal url, sitemap.instance_variable_get(:@options)[:base_url]
  end

  should 'generate sitemap index file' do
    generate_sitemap { add '/foo' }
    assert File.exists? first_sitemap_file
  end

  should 'generate static file' do
    generate_sitemap { add '/foo' }
    assert File.exists? first_sitemap_file
  end

  should 'should add paths' do
    generate_sitemap do
      add '/', {:last_modified => Time.now, :change_frequency => 'weekly', :priority => 0.5}
      add '/navigation/about/us', {:last_modified => Time.now, :change_frequency => 'weekly', :priority => 0.5}
    end

    elems = elements first_sitemap_file, 'loc'
    assert_equal 'http://example.com/', elems.first.text
    assert_equal 'http://example.com/navigation/about/us', elems.last.text
  end

  context 'Sitemap index file' do
    should 'contain one sitemapindex element' do
      generate_sitemap { add '/' }
      assert_equal 1, num_elements(sitemaps_index_file, 'sitemapindex')
    end

    should 'contain one sitemap element' do
      generate_sitemap { add '/' }
      assert_equal 1, num_elements(sitemaps_index_file, 'sitemap')
    end

    should 'contain one loc element' do
      generate_sitemap { add '/' }
      assert_equal 1, num_elements(sitemaps_index_file, 'loc')
    end

    should 'contain one lastmod element' do
      generate_sitemap { add '/' }
      assert_equal 1, num_elements(sitemaps_index_file, 'lastmod')
    end

    should 'contain two loc elements' do
      generate_sitemap(:max_per_sitemap => 2) do
        4.times { add '/' }
      end

      assert_equal 2, num_elements(sitemaps_index_file, 'loc')
    end

    should 'contain two lastmod elements' do
      generate_sitemap(:max_per_sitemap => 2) do
        4.times { add '/' }
      end

      assert_equal 2, num_elements(sitemaps_index_file, 'lastmod')
    end

    should 'not be gzipped' do
      generate_sitemap(:gzip => false) { add '/' }
      assert File.exists?(unzipped_sitemaps_index_file)
    end
  end

  context 'Sitemap file' do
    should 'contain one urlset element' do
      generate_sitemap { add '/' }
      assert_equal 1, num_elements(first_sitemap_file, 'urlset')
    end

    should 'contain several loc elements' do
      generate_sitemap do
        3.times { add '/' }
      end

      assert_equal 3, num_elements(first_sitemap_file, 'loc')
    end

    should 'contain several lastmod elements' do
      generate_sitemap do
        3.times { add '/', :last_modified => Time.now }
      end

      assert_equal 3, num_elements(first_sitemap_file, 'lastmod')
    end

    should 'contain several changefreq elements' do
      generate_sitemap do
        3.times { add '/' }
      end

      assert_equal 3, num_elements(first_sitemap_file, 'changefreq')
    end

    should 'contain several priority elements' do
      generate_sitemap do
        3.times { add '/', :priority => 0.2 }
      end

      assert_equal 3, num_elements(first_sitemap_file, 'priority')
    end

    should 'have a change frequency of weekly by default' do
      generate_sitemap do
        3.times { add '/' }
      end

      assert_equal 'weekly', elements(first_sitemap_file, 'changefreq').first.text
    end

    should 'have a change frequency of daily' do
      generate_sitemap { add '/', :change_frequency => 'daily' }
      assert_equal 'daily', elements(first_sitemap_file, 'changefreq').first.text
    end

    should 'have a priority of 0.2' do
      generate_sitemap { add '/', :priority => 0.2 }
      assert_equal '0.2', elements(first_sitemap_file, 'priority').first.text
    end

    should 'contain two loc element' do
      generate_sitemap(:max_per_sitemap => 2) do
        4.times { add '/' }
      end

      assert_equal 2, num_elements(first_sitemap_file, 'loc')
      assert_equal 2, num_elements(second_sitemap_file, 'loc')
    end

    should 'contain two changefreq elements' do
      generate_sitemap(:max_per_sitemap => 2) do
        4.times { add '/' }
      end

      assert_equal 2, num_elements(first_sitemap_file, 'changefreq')
      assert_equal 2, num_elements(second_sitemap_file, 'changefreq')
    end

    should 'contain two priority element' do
      generate_sitemap(:max_per_sitemap => 2) do
        4.times { add '/', :priority => 0.2 }
      end

      assert_equal 2, num_elements(first_sitemap_file, 'priority')
      assert_equal 2, num_elements(second_sitemap_file, 'priority')
    end

    should 'allow adding video elements' do
      video_title = "Maru Motel"
      video_url   = "http://www.youtube.com/embed/cvnPBZbth3E?rel=0"
      video_date  = Time.mktime(2010, 8, 29)

      generate_sitemap do
        add '/', :video => { :title            => video_title,
                             :player_loc       => { :value       => video_url,
                                                    :allow_embed => true },
                             :publication_date => video_date
                           }
      end

      assert_equal video_title, elements(first_sitemap_file, "*[name()='video:title']").first.text
      assert_equal video_url,   elements(first_sitemap_file, "*[name()='video:player_loc']").first.text
      assert_equal "true",      elements(first_sitemap_file, "*[name()='video:player_loc']").first.attributes["allow_embed"].value
    end

    should 'not be gzipped' do
      generate_sitemap(:gzip => false) { add '/' }
      assert File.exists?(unzipped_first_sitemap_file)
    end
  end

  context 'sanatize XML chars' do
    should 'should transform ampersands' do
      generate_sitemap { add '/something&else' }
      elems = elements(first_sitemap_file, 'loc')

      assert Zlib::GzipReader.open(first_sitemap_file).read.include?("/something&amp;else")
      assert_equal 'http://example.com/something&else', elems.first.text
    end
  end

  context 'clean method' do
    should 'be chainable' do
      sitemap = generate_sitemap { add '/' }
      assert_equal BigSitemap, sitemap.clean.class
    end

    should 'clean all sitemap files' do
      sitemap = generate_sitemap { add '/' }
      assert Dir["#{sitemaps_dir}/sitemap*"].size > 0, "#{sitemaps_dir} has sitemap files"
      sitemap.clean
      assert_equal 0, Dir["#{sitemaps_dir}/sitemap*"].size, "#{sitemaps_dir} is empty of sitemap files"
    end
  end

  context 'sitemap index' do
    should 'generate for all xml files in directory' do
      sitemap = generate_sitemap {}
      File.open("#{sitemaps_dir}/sitemap_file1.xml", 'w')
      File.open("#{sitemaps_dir}/sitemap_file2.xml.gz", 'w')
      File.open("#{sitemaps_dir}/sitemap_file3.txt", 'w')
      File.open("#{sitemaps_dir}/file4.xml", 'w')
      File.open(unzipped_sitemaps_index_file, 'w')
      sitemap.send :generate_sitemap_index

      elem = elements(sitemaps_index_file, 'loc')
      assert_equal 2, elem.size #no index and file3 and file4 found
      assert_equal "http://example.com/sitemap_file1.xml", elem.first.text
      assert_equal "http://example.com/sitemap_file2.xml.gz", elem.last.text
    end

    should 'generate for all for given file' do
      sitemap = generate_sitemap {}
      File.open("#{sitemaps_dir}/sitemap_file1.xml", 'w')
      File.open("#{sitemaps_dir}/sitemap_file2.xml.gz", 'w')
      files = ["#{sitemaps_dir}/sitemap_file1.xml", "#{sitemaps_dir}/sitemap_file2.xml.gz"]
      sitemap.send :generate_sitemap_index, files

      elem = elements(sitemaps_index_file, 'loc')
      assert_equal 2, elem.size
      assert_equal "http://example.com/sitemap_file1.xml", elem.first.text
      assert_equal "http://example.com/sitemap_file2.xml.gz", elem.last.text
    end
  end

  context 'partial update' do
    should 'not recreate old files' do
      # The first run should generate all the files
      generate_sitemap(:max_per_sitemap => 2, :partial_update => true, :gzip => false) do
        [10, 20, 30, 40, 50].each do |i|
          add "/#{i}", :id => i
        end
      end

      filename = "#{sitemaps_dir}/sitemap"

      assert File.exists? "#{filename}.xml" # ids 10 and 20
      assert File.exists? "#{filename}_30.xml" # ids 30 and 40
      assert File.exists? "#{filename}_50.xml" # id 50

      # Move the files so we can test if they are re-created
      FileUtils.mv "#{filename}.xml", "#{filename}.bak.xml"
      FileUtils.mv "#{filename}_30.xml", "#{filename}_30.bak.xml"

      # Store the original file size so we can compare it later
      original_size = File.size "#{filename}_50.xml"

      start_id = nil

      # Run a new update starting from the first ID of the last sitemap
      generate_sitemap(:max_per_sitemap => 2, :partial_update => true, :gzip => false) do
        start_id = first_id_of_last_sitemap

        [50, 60, 70, 80].each do |i|
          add "/#{i}", :id => i
        end
      end

      # Check the correct ID is returned for the beginning of the last sitemap
      assert_equal 50, start_id

      # Since we did a partial update, the earlier files shouldn't have been recreated
      assert !File.exists?("#{filename}.xml") # ids 10 and 20
      assert !File.exists?("#{filename}_30.xml") # ids 30 and 40

      # The last file of the first run should have been recreated with new records
      # and a larger file size
      assert (original_size < File.size?("#{filename}_50.xml"))

      elems = elements("#{filename}_50.xml", 'loc').map(&:text)

      assert_equal 2, elems.size

      [50, 60].each do |i|
        assert elems.include? "http://example.com/#{i}"
      end

      elems = elements(unzipped_sitemaps_index_file, 'loc').map(&:text)

      assert elems.include? 'http://example.com/sitemap.bak.xml'
      assert elems.include? 'http://example.com/sitemap_30.bak.xml'
      assert elems.include? 'http://example.com/sitemap_50.xml'
      assert elems.include? 'http://example.com/sitemap_70.xml'
    end

    context 'lockfile' do
      should 'create and delete lock file' do
        sitemap = BigSitemap.new(:base_url => 'http://example.com', :document_root => tmp_dir)

        sitemap.with_lock do
          assert File.exists?("#{sitemaps_dir}/generator.lock")
        end

        assert !File.exists?("#{sitemaps_dir}/generator.lock")
      end

      should 'not catch error not related to lock' do
        sitemap = BigSitemap.new(:base_url => 'http://example.com', :document_root => tmp_dir)

        assert_raise RuntimeError do
          sitemap.with_lock do
            raise "Wrong"
          end
        end
      end

      should 'throw error if lock exits' do
        sitemap = BigSitemap.new(:base_url => 'http://example.com', :document_root => tmp_dir)

        sitemap.with_lock do
          sitemap2 = BigSitemap.new(:base_url => 'http://example.com', :document_root => tmp_dir)

          assert_nothing_raised do
            sitemap2.with_lock do
              raise "Should not be called"
            end
          end

        end
      end
    end
  end

end


test/fixtures/test_model.rb
class TestModel
  def to_param
    id
  end

  def id
    @id ||= TestModel.current_id += 1
  end

  def change_frequency
    'monthly'
  end

  def priority
    0.8
  end

  def updated_at
    Time.at(1000000000)
  end

  class << self
    def table_name
      'test_models'
    end

    def count_for_sitemap
      self.find_for_sitemap.size
    end

    def num_items
      10
    end

    def find_for_sitemap(options={})
      instances = []
      num_times = options.delete(:limit) || self.num_items
      num_times.times { instances.push(self.new) }
      instances
    end

    attr_writer :current_id

    def current_id
      @current_id ||= 0
    end
  end
end

lib/big_sitemap.rb
require 'uri'
require 'fileutils'

require 'big_sitemap/builder'

class BigSitemap
  DEFAULTS = {
    :max_per_sitemap => Builder::MAX_URLS,
    :document_path   => '/',
    :gzip            => true,

    # Opinionated
    :ping_google => true,
    :ping_yahoo  => false, # needs :yahoo_app_id
    :ping_bing   => false,
    :ping_ask    => false, 
    :ping_yandex => false
  }

  class << self
    def generate(options={}, &block)
      @sitemap = self.new(options)

      @sitemap.first_id_of_last_sitemap = first_id_of_last_sitemap

      instance_eval(&block)

      @sitemap.with_lock do
        @sitemap.generate(options)
      end
    end

    private

    def first_id_of_last_sitemap
      Dir["#{@sitemap.document_full}sitemap*.{xml,xml.gz}"].map do |file|
        file.to_s.scan(/sitemap_(.+).xml/).flatten.last.to_i
      end.sort.last
    end

    def add(path, options={})
      @sitemap.add_path(path, options)
    end
  end

  def initialize(options={})
    @options = DEFAULTS.merge options

    if @options[:max_per_sitemap] <= 1
      raise ArgumentError, '":max_per_sitemap" must be greater than 1'
    end

    if @options[:url_options] && !@options[:base_url]
      @options[:base_url] = URI::Generic.build( {:scheme => "http"}.merge(@options.delete(:url_options)) ).to_s
    end

    unless @options[:base_url]
      raise ArgumentError, 'you must specify either ":url_options" hash or ":base_url" string'
    end
    @options[:url_path] ||= @options[:document_path]

    unless @options[:document_root]
      raise ArgumentError, 'Document root must be specified with the ":document_root" option"'
    end

    @options[:document_full] ||= File.join(@options[:document_root], @options[:document_path])
    unless @options[:document_full]
      raise ArgumentError, 'Document root must be specified with the ":document_root" option, the full path with ":document_full"'
    end

    Dir.mkdir(@options[:document_full]) unless File.exists?(@options[:document_full])

    @sources       = []
    @models        = []
    @sitemap_files = []
  end

  def first_id_of_last_sitemap
    @first_id_of_last_sitemap
  end

  def first_id_of_last_sitemap=(first_id)
    @first_id_of_last_sitemap = first_id
  end

  def document_full
    @options[:document_full]
  end

  def add(model, options={})
    warn 'BigSitemap#add is deprecated.  Please use BigSitemap.generate and call add inside the block (in BigSitemap 1.0.0+).  You will have to perform the find and generate the path for each record yourself.'
    @models << model

    filename_suffix = @models.count(model) - 1

    options[:path]           ||= table_name(model)
    options[:filename]       ||= file_name(model)
    options[:primary_column] ||= 'id' if model.new.respond_to?('id')
    options[:partial_update]   = @options[:partial_update] && options[:partial_update] != false

    options[:filename] << "_#{filename_suffix}" unless filename_suffix == 0

    @sources << [model, options.dup]

    self
  end

  def add_path(path, options)
    @paths ||= []
    @paths << [path, options]
    self
  end

  def add_static(url, time = nil, frequency = nil, priority = nil)
    warn 'BigSitemap#add_static is deprecated.  Please use BigSitemap#add_path instead'
    @static_pages ||= []
    @static_pages << [url, time, frequency, priority]
    self
  end

  def with_lock
    lock!
    begin
      yield
    ensure
      unlock!
    end
  rescue Errno::EACCES => e
    STDERR.puts 'Lockfile exists' if $VERBOSE
  end

  def file_name(name=nil)
    name   = table_name(name) unless (name.nil? || name.is_a?(String))
    prefix = 'sitemap'
    prefix << '_' unless name.nil?
    File.join(@options[:document_full], "#{prefix}#{name}")
  end

  def dir_files
    File.join(@options[:document_full], "sitemap*.{xml,xml.gz}")
  end

  def clean
    Dir[dir_files].each do |file|
      FileUtils.rm file
    end

    self
  end

  # TODO: Deprecate (move to private)
  def generate(options={})
    clean unless options[:partial_update]

    add_urls

    generate_sitemap_index

    ping_search_engines

    self
  end

  def add_urls
    return self if Array(@paths).empty?

    with_sitemap do |builder|
      @paths.uniq!
      @paths.each do |path, options|
        url = URI.join(@options[:base_url], path)
        builder.add_url! url, options
      end
    end

    self
  end

  # Create a sitemap index document
  def generate_sitemap_index(files=nil)
    files ||= Dir[dir_files]

    with_sitemap({:name => 'index', :type => 'index'}) do |sitemap|
      for path in files
        next if path =~ /index/
        sitemap.add_url! url_for_sitemap(path), :last_modified => File.stat(path).mtime
      end
    end

    self
  end

  def ping_search_engines
    require 'net/http'
    require 'cgi'

    sitemap_uri = CGI::escape(url_for_sitemap(@sitemap_files.last))

    if @options[:ping_google]
      Net::HTTP.get('www.google.com', "/webmasters/tools/ping?sitemap=#{sitemap_uri}")
    end

    if @options[:ping_yahoo]
      if @options[:yahoo_app_id]
        Net::HTTP.get(
          'search.yahooapis.com', "/SiteExplorerService/V1/updateNotification?" +
            "appid=#{@options[:yahoo_app_id]}&url=#{sitemap_uri}"
        )
      else
        STDERR.puts 'unable to ping Yahoo: no ":yahoo_app_id" provided'
      end
    end

    if @options[:ping_bing]
      Net::HTTP.get('www.bing.com', "/webmaster/ping.aspx?siteMap=#{sitemap_uri}")
    end

    if @options[:ping_ask]
      Net::HTTP.get('submissions.ask.com', "/ping?sitemap=#{sitemap_uri}")
    end

    if @options[:ping_yandex]
      Net::HTTP.get('webmaster.yandex.ru', "/wmconsole/sitemap_list.xml?host=#{sitemap_uri}")
    end
  end

  private

  def lock!(lock_file = 'generator.lock')
    lock_file = File.join(@options[:document_full], lock_file)
    File.open(lock_file, 'w', File::EXCL)
  end

  def unlock!(lock_file = 'generator.lock')
    lock_file = File.join(@options[:document_full], lock_file)
    FileUtils.rm lock_file
  end

  def with_sitemap(options={})
    options[:filename]       ||= file_name(options[:name])
    options[:type]           ||= 'sitemap'
    options[:max_urls]       ||= @options["max_per_#{options[:type]}".to_sym]
    options[:gzip]           ||= @options[:gzip]
    options[:indent]         ||= 2
    options[:partial_update] ||= @options[:partial_update]
    options[:start_part_id]  ||= first_id_of_last_sitemap

    sitemap = if options[:type] == 'index'
      IndexBuilder.new(options)
    else
      Builder.new(options)
    end

    begin
      yield sitemap
    ensure
      sitemap.close!
      @sitemap_files.concat sitemap.filepaths!
    end
  end

  def pick_method(model, candidates)
    method = nil
    candidates.each do |candidate|
      if model.respond_to? candidate
        method = candidate
        break
      end
    end
    method
  end

  def url_for_sitemap(path)
    File.join @options[:base_url], @options[:url_path], File.basename(path)
  end

end


class BigSitemapRails < BigSitemap
  def self.generate(options={}, &block)
    raise 'No Rails Environment loaded' unless defined? Rails

    DEFAULTS.merge!(:document_root => "#{Rails.root}/public", :url_options => default_url_options)
    super(options, &block)
  end
end


class BigSitemapMerb < BigSitemap
  def self.generate(options={}, &block)
    raise 'No Merb Environment loaded' unless defined? ::Merb
    require 'extlib'

    DEFAULTS.merge!(:document_root => "#{Merb.root}/public")
    super(options, &block)
  end
end


lib/big_sitemap/builder.rb
require 'fileutils'
require 'zlib'

class BigSitemap
  class Builder
    MAX_URLS = 50000
    HEADER_NAME = 'urlset'
    HEADER_ATTRIBUTES = {
      'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9',
      'xmlns:video' => "http://www.google.com/schemas/sitemap-video/1.1",
      'xmlns:xsi' => "http://www.w3.org/2001/XMLSchema-instance",
      'xsi:schemaLocation' => "http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"
    }

    # as per http://support.google.com/webmasters/bin/answer.py?hl=en&answer=80472#1
    VIDEO_ATTRIBUTES = %w(thumbnail_loc title description content_loc player_loc duration expiration_date rating view_count
                          publication_date family_friendly restriction gallery_loc price requires_subscription uploader platform live)

    def initialize(options)
      @gzip           = options.delete(:gzip)
      @max_urls       = options.delete(:max_urls) || MAX_URLS
      @type           = options.delete(:type)
      @filepaths      = []
      @parts          = options.delete(:start_part_id) || 0
      @partial_update = options.delete(:partial_update)

      @filename         = options.delete(:filename)
      @current_filename = nil
      @tmp_filename     = nil
      @target           = _get_writer

      @level = 0
      @opened_tags = []
      _init_document
    end

    def add_url!(location, options={})
      _rotate(options[:id]) if @max_urls == @urls
      _open_tag 'url'

      tag! 'loc', location
      tag! 'lastmod', options[:last_modified].utc.strftime('%Y-%m-%dT%H:%M:%S+00:00') if options[:last_modified]
      tag! 'changefreq', options[:change_frequency] || 'weekly'
      tag! 'priority', options[:priority] if options[:priority]

      if options[:video]
        _open_tag 'video:video'

          options[:video].each do |attribute, value_or_hash|
            if value_or_hash.is_a?(Hash)
              tag_value = value_or_hash.delete(:value)
              opts      = value_or_hash
            else
              tag_value = value_or_hash
              opts      = {}
            end

            tag_value = tag_value.utc.strftime('%Y-%m-%dT%H:%M:%S+00:00') if attribute.to_s[0..-5] == "_date"

            tag! "video:#{attribute}", tag_value, opts
          end

        _close_tag 'video:video'
      end

      _close_tag 'url'

      @urls += 1
    end

    def filepaths!
      @filepaths
    end

    def close!
      _close_document
      target!.close if target!.respond_to?(:close)
      File.delete(@current_filename) if File.exists?(@current_filename)
      File.rename(@tmp_filename, @current_filename)
    end

    def target!
      @target
    end

    private

    def _get_writer
      filename = @filename.dup
      filename << "_#{@parts}" if @parts > 0 && @type != 'index'
      filename << '.xml'
      filename << '.gz' if @gzip
      _open_writer(filename)
    end

    def _open_writer(filename)
      @current_filename = filename
      @tmp_filename     = filename + ".tmp"
      @filepaths << filename
      file = ::File.open(@tmp_filename, 'w+:ASCII-8BIT')
      @gzip ? ::Zlib::GzipWriter.new(file) : file
    end

    def _init_document
      @urls = 0
      target!.print '<?xml version="1.0" encoding="UTF-8"?>'
      _newline
      _open_tag self.class::HEADER_NAME, self.class::HEADER_ATTRIBUTES
    end

    def _rotate(part_nr=nil)
      # write out the current document and start writing into a new file
      close!
      @parts = part_nr || @parts + 1
      @target = _get_writer
      _init_document
    end

    # opens a tag, bumps up level but doesn't require a block
    def _open_tag(name, attrs={})
      _indent
      _start_tag(name, attrs)
      _newline
      @level += 1
      @opened_tags << name
    end

    def _start_tag(name, attrs={})
      attrs = attrs.map { |attr, value| %Q( #{attr}="#{value}") }.join('')
      target!.print "<#{name}#{attrs}>"
    end

    def tag!(name, content, attrs = {})
      _indent
      _start_tag(name, attrs)
      target!.print content.to_s.gsub('&', '&amp;')
      _end_tag(name)
      _newline
    end

    def _end_tag(name)
      target!.print "</#{name}>"
    end

    # closes a tag block by decreasing the level and inserting a close tag
    def _close_tag(name)
      @opened_tags.pop
      @level -= 1
      _indent
      _end_tag(name)
      _newline
    end

    def _close_document
      for name in @opened_tags.reverse
        _close_tag(name)
      end
    end

    def _indent
      target!.print "  " * @level
    end

    def _newline
      target!.puts ''
    end
  end

  class IndexBuilder < Builder
    HEADER_NAME = 'sitemapindex'
    HEADER_ATTRIBUTES = {
      'xmlns' => 'http://www.sitemaps.org/schemas/sitemap/0.9'
    }

    def add_url!(location, options={})
      _open_tag 'sitemap'

      tag! 'loc', location
      tag! 'lastmod', options[:last_modified].utc.strftime('%Y-%m-%dT%H:%M:%S+00:00') if options[:last_modified]

      _close_tag 'sitemap'
    end
  end


end
